File: game.py
from cards import Cards
from random import seed

def play(players) -> int:
    """
    Plays the game with the given list of players until one
    player wins or there is a draw. If a player wins, the
    function returns the number of the player (0 to one less
    than the number of players). If there is a draw, the function
    returns -1.
    """
    number_of_players = len(players)
    cards = Cards(number_of_players)
    history = set()
    while True:
        for i, p in enumerate(players):
            cards.show(i)
            if cards.is_empty(i):
                print(f"Player {i} must skip as they have no cards")
                continue

            other, suit = p.next_move(i, cards, history)
            print(f"Player {i} requests suit {suit} from player {other}")
            if players[other].has_card(other, i, suit, cards, history):
                print(f"Player {other} hands card {suit} to player {i}")
                cards.transfer(suit, other, i, False)
            else:
                print(f"Player {other} has no cards of suit {suit}")
                cards.no_transfer(suit, other, i, False)
            winner = cards.test_winner(i)
            if winner >= 0:
                cards.show(-1)
                return winner

            # if a position repeats, it forces a draw
            position = cards.position(i)
            if position in history:
                return -1
            history.add(position)

File: src/game.rs
use cards::{Cards, NO_WINNER, ILLEGAL_CARDS};
use player::Player;
use std::collections::HashSet;

/** 
    Plays the game with the given list of players until one
    player wins or there is a draw. If a player wins, the
    function returns the number of the player (0 to one less
    than the number of players). If there is a draw, the function
    returns -1.
*/
pub fn play(players: &[usize], player_instances: &mut [Box<Player>]) -> i64 {
    let number_of_players = players.len();
    assert!(player_instances.len() <= number_of_players);

    let mut cards = Cards::new(number_of_players);
    let mut history = HashSet::new();

    loop {
        for i in 0..number_of_players {
            cards.show(i);
            if cards.is_empty(i) {
                println!("Player {} must skip as they have no cards", i);
                continue;
            }
            let (other, suit) = player_instances[players[i]].next_move(i, &cards, &history);
            println!("Player {} requests suit {} from player {}", i, suit, other);
            if player_instances[players[other]].has_card(other, i, suit, &cards, &history) {
                println!("Player {} hands card {} to player {}", suit, other, i);
                cards.transfer(suit, other, i, false);
            } else {
                println!("Player {} has no cards of suit {}", other, suit);
                cards.no_transfer(suit, other, i, false);
            }
            let winner = cards.test_winner(i);
            if winner == ILLEGAL_CARDS {
                cards.show(usize::max_value());
                panic!("The cards are in an illegal state. All players lose");
            }
            if winner != NO_WINNER {
                cards.show(usize::max_value());
                return winner
            }
            let position = cards.position(i);
            if history.contains(&position) {
                cards.show(usize::max_value());
                return -1
            }
            history.insert(position);
        }
    }
}

